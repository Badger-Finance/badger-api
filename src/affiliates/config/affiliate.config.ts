import { BadRequest } from '@tsed/exceptions';
import { Chain } from '../../chains/config/chain.config';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';

type Affiliates = { [protocol: string]: Affiliate };

export abstract class Affiliate {
  private static affiliates: Affiliates = {};

  readonly name: string;

  constructor(name: string) {
    this.name = name;
  }

  static register(protocol: Protocol, affiliate: Affiliate): void {
    this.affiliates[protocol] = affiliate;
  }

  static getAffiliate(affiliateData: SettAffiliateData): Affiliate {
    const affiliate = this.affiliates[affiliateData.protocol];
    if (!affiliate) {
      throw new BadRequest(`${affiliate} is not a affiliate`);
    }
    return affiliate;
  }

  abstract getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData>;
}
