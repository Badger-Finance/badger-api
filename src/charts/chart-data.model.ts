import { attribute } from '@aws/dynamodb-data-mapper-annotations';

export abstract class ChartData<T> {
  @attribute()
  id!: string;

  @attribute()
  timestamp!: number;

  abstract toBlankData(): T;
}
