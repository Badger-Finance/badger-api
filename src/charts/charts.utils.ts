import { ChartTimeFrame, ONE_DAY_MS, ONE_HOUR_MS } from '@badger-dao/sdk';

import { getDataMapper } from '../aws/dynamodb.utils';
import { ChartDataBlob } from '../aws/models/chart-data-blob.model';
import { HistoricVaultSnapshotModel } from '../aws/models/historic-vault-snapshot.model';
import { ChartData } from './chart-data.model';

// list of ChartTimeFrame enums that contain unique capture granularities for searching
export const CHART_GRANULARITY_TIMEFRAMES = [ChartTimeFrame.Max, ChartTimeFrame.Week, ChartTimeFrame.Day];

/**
 *
 * @param id
 * @param timeframe
 * @param data
 * @returns
 */
export function toChartDataBlob<T extends ChartData<T>>(
  id: string,
  timeframe: ChartTimeFrame,
  data: T[]
): ChartDataBlob<T> {
  return Object.assign(new ChartDataBlob(), {
    id,
    timeframe,
    data
  });
}

/**
 *
 * @param namespace
 * @param id
 * @param timeframe
 * @returns
 */
export function toChartDataKey(namespace: string, id: string, timeframe: ChartTimeFrame): string {
  return [namespace, id, timeframe].join('_');
}

/**
 *
 * @param reference
 * @param timestamp
 * @param timeframe
 * @returns
 */
export function shouldUpdate(reference: number, timestamp: number, timeframe: ChartTimeFrame): boolean {
  const difference = reference - timestamp;
  let update = false;
  switch (timeframe) {
    case ChartTimeFrame.Max:
    case ChartTimeFrame.Year:
    case ChartTimeFrame.ThreeMonth:
    case ChartTimeFrame.Month:
    case ChartTimeFrame.YTD:
      update = difference >= ONE_DAY_MS;
      break;
    case ChartTimeFrame.Week:
      update = difference >= ONE_HOUR_MS * 6;
      break;
    default:
      update = difference >= ONE_HOUR_MS;
  }
  return update;
}

/**
 *
 * @param reference
 * @param timestamp
 * @param timeframe
 * @returns
 */
export function shouldTrim(reference: number, timestamp: number, timeframe: ChartTimeFrame): boolean {
  const difference = reference - timestamp;
  let update;
  switch (timeframe) {
    case ChartTimeFrame.Year:
      update = difference >= ONE_DAY_MS * 365;
      break;
    case ChartTimeFrame.ThreeMonth:
      update = difference >= ONE_DAY_MS * 90;
      break;
    case ChartTimeFrame.Month:
      update = difference >= ONE_DAY_MS * 30;
      break;
    case ChartTimeFrame.Week:
      update = difference >= ONE_DAY_MS * 7;
      break;
    case ChartTimeFrame.Day:
      update = difference >= ONE_DAY_MS;
      break;
    // Year to date needs a full clear - so we will never trim
    case ChartTimeFrame.YTD:
    default:
      update = false;
  }
  return update;
}

/**
 *
 * @param namespace
 * @param snapshot
 */
export async function updateSnapshots<T extends ChartData<T>>(namespace: string, snapshot: T) {
  const mapper = getDataMapper();
  const isFirstOfYear = (date: Date) => date.getDay() === 0 && date.getMonth() === 0;

  const { id, timestamp: now } = snapshot;
  for (const timeframe of Object.values(ChartTimeFrame)) {
    const searchKey = Object.assign(new ChartDataBlob<T>(), {
      id: toChartDataKey(namespace, id, timeframe)
    });

    let cachedChart: ChartDataBlob<T> | undefined;
    try {
      cachedChart = await mapper.get(searchKey);
      if (timeframe === ChartTimeFrame.YTD) {
        const date = new Date(now);
        if (isFirstOfYear(date) && cachedChart.data.length > 1) {
          // new year, force a new object to be created
          cachedChart = undefined;
        }
      }
    } catch (err) {
      console.debug({ message: 'Unable to query cached chart, may simply not exist', err });
    } // no item found

    let updateCache = false;
    if (!cachedChart) {
      const blob = toChartDataBlob(searchKey.id, timeframe, []);
      console.debug(`Create blob for ${searchKey.id}`);
      try {
        cachedChart = await mapper.put(blob);
      } catch (err) {
        console.error({ message: 'Unable to save blob', err });
      }
      updateCache = true;
    } else {
      const { timeframe, data } = cachedChart;
      const recentSnapshot = data[0].timestamp;
      updateCache = shouldUpdate(now, recentSnapshot, timeframe);

      const lastSnapshot = data[data.length - 1].timestamp;
      if (shouldTrim(now, lastSnapshot, timeframe)) {
        cachedChart.data = cachedChart.data.slice(0, data.length - 1);
      }
    }

    if (updateCache && cachedChart) {
      cachedChart.data.splice(0, 0, snapshot);
      console.debug(`Update ${searchKey.id} (${cachedChart.data.length} entries)`);
      try {
        await mapper.put(cachedChart);
      } catch (err) {
        console.error({ message: 'Unable to save blob', err });
      }
    }
  }
}

export async function queryVaultCharts(id: string): Promise<HistoricVaultSnapshotModel[]> {
  try {
    const mapper = getDataMapper();
    for await (const item of mapper.query(ChartDataBlob, { id }, { limit: 1, scanIndexForward: false })) {
      return item.data as HistoricVaultSnapshotModel[];
    }
    return [];
  } catch (err) {
    console.error(err);
    return [];
  }
}
