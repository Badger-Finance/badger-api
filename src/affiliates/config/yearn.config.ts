import { BigNumber, ethers } from 'ethers';
import { Chain } from '../../chains/config/chain.config';
import { guestListAbi } from '../../config/abi/guest-list.abi';
import { yearnAffiliateVaultWrapperAbi } from '../../config/abi/yearn-affiliate-vault-wrapper.abi';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { getToken } from '../../tokens/tokens.utils';
import { Affiliate } from './affiliate.config';

export class Yearn extends Affiliate {
  constructor() {
    super(Protocol.Yearn);
    Affiliate.register(Protocol.Yearn, this);
  }

  async getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData> {
    const affiliateData: SettAffiliateData = {
      protocol: Protocol.Yearn,
    };
    if (sett.hasBouncer) {
      try {
        const vaultWrapper = new ethers.Contract(sett.settToken, yearnAffiliateVaultWrapperAbi, chain.provider);
        const guestListAddress = await vaultWrapper.guestList();
        const guesList = new ethers.Contract(guestListAddress, guestListAbi, chain.provider);
        const [available, limit]: [BigNumber, BigNumber] = await Promise.all([
          guesList.remainingTotalDepositAllowed(),
          guesList.totalDepositCap(),
        ]);
        const depositToken = getToken(sett.depositToken);
        affiliateData.availableDepositLimit = parseFloat(ethers.utils.formatUnits(available, depositToken.decimals));
        affiliateData.depositLimit = parseFloat(ethers.utils.formatUnits(limit, depositToken.decimals));
      } catch {
        affiliateData.availableDepositLimit = 0;
        affiliateData.depositLimit = 50;
      }
    }
    return affiliateData;
  }
}
