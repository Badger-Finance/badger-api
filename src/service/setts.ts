import { TOKENS } from '../config/constants';

export type SettData = {
	name: string;
	symbol: string;
	depositToken: string;
	settToken: string;
	geyserAddress?: string;
	protocol?: string;
};

export const setts: SettData[] = [
	{
		name: 'Curve.fi renBTC/wBTC/sBTC',
		symbol: 'sBTCCRV',
		depositToken: TOKENS.CRV_SBTC,
		settToken: '0xd04c48a53c111300ad41190d63681ed3dad998ec',
		geyserAddress: '0x10fc82867013fce1bd624fafc719bb92df3172fc',
		protocol: 'curve',
	},
	{
		name: 'Curve.fi crvRenWBTC',
		symbol: 'renBTCCRV',
		depositToken: TOKENS.CRV_RENBTC,
		settToken: '0x6def55d2e18486b9ddfaa075bc4e4ee0b28c1545',
		geyserAddress: '0x2296f174374508278dc12b806a7f27c87d53ca15',
		protocol: 'curve',
	},
	{
		name: 'Curve.fi tBTC/sBTCCrv LP',
		symbol: 'tBTCCRV',
		depositToken: TOKENS.CRV_TBTC,
		settToken: '0xb9d076fde463dbc9f915e5392f807315bf940334',
		geyserAddress: '0x085a9340ff7692ab6703f17ab5ffc917b580a6fd',
		protocol: 'curve',
	},
	{
		name: 'Harvest Curve.fi crvRenWBTC',
		symbol: 'hrenBTCCRV',
		depositToken: TOKENS.CRV_RENBTC,
		settToken: '0xaf5a1decfa95baf63e0084a35c62592b774a2a87',
		geyserAddress: '0xed0b7f5d9f6286d00763b0ffcba886d8f9d56d5e',
		protocol: 'curve',
	},
	{
		name: 'Badger',
		symbol: 'BADGER',
		depositToken: TOKENS.BADGER,
		settToken: '0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28',
		geyserAddress: '0xa9429271a28f8543efffa136994c0839e7d7bf77',
	},
	{
		name: 'Uniswap Wrapped BTC/Badger',
		symbol: 'BADGER-WBTC',
		depositToken: TOKENS.UNI_BADGER_WBTC,
		settToken: '0x235c9e24d3fb2fafd58a2e49d454fdcd2dbf7ff1',
		geyserAddress: '0xa207d69ea6fb967e54baa8639c408c31767ba62d',
		protocol: 'uniswap',
	},
	{
		name: 'Sushiswap Wrapped BTC/Wrapped Ether',
		symbol: 'SLP-WBTC-ETH',
		depositToken: TOKENS.SUSHI_ETH_WBTC,
		settToken: '0x758a43ee2bff8230eeb784879cdcff4828f2544d',
		geyserAddress: '0x612f681bcd12a0b284518d42d2dbcc73b146eb65',
		protocol: 'sushiswap',
	},
	{
		name: 'Sushiswap Wrapped BTC/Badger',
		symbol: 'SLP-BADGER-WBTC',
		depositToken: TOKENS.SUSHI_BADGER_WBTC,
		settToken: '0x1862a18181346ebd9edaf800804f89190def24a5',
		geyserAddress: '0xb5b654efba23596ed49fade44f7e67e23d6712e7',
		protocol: 'sushiswap',
	},
	{
		name: 'Digg',
		symbol: 'DIGG',
		depositToken: TOKENS.DIGG,
		settToken: '0x7e7e112a68d8d2e221e11047a72ffc1065c38e1a',
	},
	{
		name: 'Uniswap Wrapped BTC/Digg',
		symbol: 'DIGG-WBTC',
		depositToken: TOKENS.UNI_DIGG_WBTC,
		settToken: '0xc17078fdd324cc473f8175dc5290fae5f2e84714',
		geyserAddress: '0x0194b5fe9ab7e0c43a08acbb771516fc057402e7',
		protocol: 'uniswap',
	},
	{
		name: 'Sushiswap Wrapped BTC/Digg',
		symbol: 'SLP-DIGG-WBTC',
		depositToken: TOKENS.SUSHI_DIGG_WBTC,
		settToken: '0x88128580acdd9c04ce47afce196875747bf2a9f6',
		geyserAddress: '0x7f6fe274e172ac7d096a7b214c78584d99ca988b',
		protocol: 'sushiswap',
	},
];
