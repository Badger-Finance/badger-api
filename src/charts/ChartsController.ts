import { Controller, Get, Inject, QueryParams } from '@tsed/common';
import { ContentType } from '@tsed/schema';
import { SettSnapshot } from '../interface/SettSnapshot';
import { SettsService } from '../setts/SettsService';

@Controller('/charts')
export class ChartsController {
  @Inject()
  settsService!: SettsService;

  @Get()
  @ContentType('json')
  getChart(@QueryParams('settName') settName: string, @QueryParams('count') count?: number): Promise<SettSnapshot[]> {
    return this.settsService.getSettSnapshots(settName, count);
  }
}
