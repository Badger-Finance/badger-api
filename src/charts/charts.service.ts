import { Service } from '@tsed/di';
import { Chain } from '../chains/config/chain.config';
import { SettDefinition } from '../setts/interfaces/sett-definition.interface';
import { SettSnapshot } from '../setts/interfaces/sett-snapshot.interface';
import { getSettSnapshotsInRange } from '../setts/setts.utils';
import { ChartGranularity } from './enums/chart-granularity.enum';

@Service()
export class ChartsService {
  async getChartData(
    chain: Chain,
    sett: SettDefinition,
    start: Date,
    end: Date,
    granularity: ChartGranularity,
    period: number,
  ): Promise<SettSnapshot[]> {
    let snapshots = await getSettSnapshotsInRange(sett, start, end);

    // snapshot granularity @ 30 min intervals, 2 per hour, 48 per day
    const interval = granularity === ChartGranularity.HOUR ? 2 : 48 * period;
    // filter data down to selected granularity period combination
    snapshots = snapshots.filter((_s, i) => i % interval === 0);

    const expectedDataPoints = this.getRequestedDataPoints(start, end, granularity, period);
    if (snapshots.length < expectedDataPoints) {
      let lastSnapshot = snapshots[snapshots.length - 1];
      if (!lastSnapshot) {
        lastSnapshot = {
          address: sett.settToken,
          value: 0,
          balance: 0,
          supply: 0,
          ratio: 1,
          height: await chain.provider.getBlockNumber(),
          timestamp: Date.now(),
        };
      }
      const thirtyMinutesBlocks = parseInt((chain.blocksPerYear / 365 / 24 / 2).toString());
      while (snapshots.length < expectedDataPoints) {
        const newSnapshot: SettSnapshot = {
          address: sett.settToken,
          value: 0,
          balance: 0,
          supply: 0,
          ratio: 1,
          height: lastSnapshot.height - thirtyMinutesBlocks * (interval ?? 1),
          timestamp: lastSnapshot.timestamp - 30 * 60 * 1000 * (interval ?? 1),
        };
        snapshots.push(newSnapshot);
        lastSnapshot = newSnapshot;
      }
    }

    return snapshots;
  }

  getRequestedDataPoints(start: Date, end: Date, granularity: ChartGranularity, period: number): number {
    let returnedDataPoints: number;
    const diffInMs = new Date(end).getTime() - new Date(start).getTime();
    if (granularity === ChartGranularity.DAY) {
      const diffInDays = diffInMs / (1000 * 60 * 60 * 24);
      returnedDataPoints = Math.floor(diffInDays / period);
    } else {
      const diffInHours = diffInMs / (1000 * 60 * 60);
      returnedDataPoints = Math.floor(diffInHours / period);
    }
    return returnedDataPoints;
  }
}
