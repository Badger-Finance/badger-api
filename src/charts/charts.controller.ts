import { Controller, Get, Inject, QueryParams, UsePipe } from '@tsed/common';
import { NotFound, UnprocessableEntity } from '@tsed/exceptions';
import { ContentType, Hidden } from '@tsed/schema';
import { ethers } from 'ethers';
import { Chain } from '../chains/config/chain.config';
import { ValidationPipe } from '../common/decorators/validation-pipe';
import { VaultSnapshot } from '../vaults/interfaces/vault-snapshot.interface';
import { ChartsService } from './charts.service';
import { ChartsQueryDto } from './dto/charts-query.dto';
import { ChartGranularity } from './enums/chart-granularity.enum';

@Controller('/charts')
export class ChartsController {
  @Inject()
  chartsService!: ChartsService;

  @Hidden()
  @Get()
  @ContentType('json')
  async getChart(
    @UsePipe(ValidationPipe)
    @QueryParams()
    query: ChartsQueryDto,
  ): Promise<VaultSnapshot[]> {
    const now = new Date();
    const yesterday = new Date();
    yesterday.setHours(now.getHours() - 24);

    const {
      id: settToken,
      chain,
      start = yesterday,
      end = now,
      period = 1,
      granularity = ChartGranularity.HOUR,
    } = query;

    const { isValid, error } = this.isValidGranularityPeriod(start, end, granularity, period);

    if (!isValid && error) {
      throw new UnprocessableEntity(error);
    }

    let checksumContract = settToken;
    try {
      checksumContract = ethers.utils.getAddress(settToken);
    } catch (err) {
      throw new UnprocessableEntity('Invalid contract address');
    }

    const vault = Chain.getChain(chain).vaults.find((vault) => vault.vaultToken === checksumContract);

    if (!vault) {
      throw new NotFound(`${checksumContract} is not a valid sett`);
    }

    return this.chartsService.getChartData(vault, start, end, granularity, period);
  }

  private isValidGranularityPeriod(
    start: Date,
    end: Date,
    granularity: ChartGranularity,
    period: number,
  ): { isValid: boolean; error?: string } {
    const maxDataPointsToReturn = 50;

    if (start >= end) {
      return {
        isValid: false,
        error: '"start" must be strictly less than "end"',
      };
    }

    const returnedDataPoints = ChartsService.getRequestedDataPoints(start, end, granularity, period);
    if (returnedDataPoints > maxDataPointsToReturn) {
      return {
        isValid: false,
        error: 'Too many data points to return. Increase the period or shorten the time range.',
      };
    }

    return {
      isValid: true,
    };
  }
}
