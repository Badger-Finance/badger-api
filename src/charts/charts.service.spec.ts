import { PlatformTest } from '@tsed/common';
import { Ethereum } from '../chains/config/eth.config';
import { TOKENS } from '../config/tokens.config';
import { getSettDefinition } from '../setts/setts.utils';
import { randomSnapshots, setupMapper } from '../test/tests.utils';
import { ChartsService } from './charts.service';
import { ChartGranularity } from './enums/chart-granularity.enum';

describe('charts.service', () => {
  let service: ChartsService;

  beforeAll(async () => {
    await PlatformTest.create();

    service = PlatformTest.get<ChartsService>(ChartsService);
  });

  afterEach(PlatformTest.reset);

  describe('getRequestedDataPoints', () => {
    const today = new Date();
    const yesterday = new Date();
    yesterday.setDate(yesterday.getDate() - 1);
    const oneWeek = new Date();
    oneWeek.setDate(oneWeek.getDate() - 7);
    const oneMonth = new Date();
    oneMonth.setDate(oneMonth.getDate() - 30);
    it.each([
      [24, 1, ChartGranularity.HOUR, yesterday, today],
      [12, 2, ChartGranularity.HOUR, yesterday, today],
      [1, 24, ChartGranularity.HOUR, yesterday, today],
      [1, 1, ChartGranularity.DAY, yesterday, today],
      [7, 1, ChartGranularity.DAY, oneWeek, today],
      [7 * 24, 1, ChartGranularity.HOUR, oneWeek, today],
      [7 * 12, 2, ChartGranularity.HOUR, oneWeek, today],
      [30, 1, ChartGranularity.DAY, oneMonth, today],
      [60, 12, ChartGranularity.HOUR, oneMonth, today],
      [0, 12, ChartGranularity.HOUR, today, today],
      [0, 1, ChartGranularity.HOUR, today, today],
      [0, 1, ChartGranularity.DAY, today, today],
      [0, 12, ChartGranularity.DAY, today, today],
    ])(
      'returns %d data points for granularity %d %s from %s to %s',
      (count: number, size: number, granularity: ChartGranularity, start: Date, end: Date) => {
        const actual = ChartsService.getRequestedDataPoints(new Date(start), new Date(end), granularity, size);
        expect(actual).toEqual(count);
      },
    );
  });

  describe('getChartData', () => {
    const today = new Date();
    const yesterday = new Date();
    yesterday.setDate(yesterday.getDate() - 1);
    const oneWeek = new Date();
    oneWeek.setDate(oneWeek.getDate() - 7);
    const oneMonth = new Date();
    oneMonth.setDate(oneMonth.getDate() - 30);
    it.each([
      [24, 1, ChartGranularity.HOUR, yesterday, today],
      [12, 2, ChartGranularity.HOUR, yesterday, today],
      [1, 24, ChartGranularity.HOUR, yesterday, today],
      [1, 1, ChartGranularity.DAY, yesterday, today],
      [7, 1, ChartGranularity.DAY, oneWeek, today],
      [7 * 24, 1, ChartGranularity.HOUR, oneWeek, today],
      [7 * 12, 2, ChartGranularity.HOUR, oneWeek, today],
      [30, 1, ChartGranularity.DAY, oneMonth, today],
      [60, 12, ChartGranularity.HOUR, oneMonth, today],
      [0, 12, ChartGranularity.HOUR, today, today],
      [0, 1, ChartGranularity.HOUR, today, today],
      [0, 1, ChartGranularity.DAY, today, today],
      [0, 12, ChartGranularity.DAY, today, today],
    ])(
      'returns %d data points for granularity %d %s from %s to %s',
      async (count: number, size: number, granularity: ChartGranularity, start: Date, end: Date) => {
        const expected = ChartsService.getRequestedDataPoints(new Date(start), new Date(end), granularity, size);
        const settDefinition = getSettDefinition(new Ethereum(), TOKENS.BBADGER);
        // snapshot granularity @ 30 min intervals, 2 per hour, 48 per day
        const interval = granularity === ChartGranularity.HOUR ? 2 : 48 * size;
        const seed = randomSnapshots(settDefinition, count * interval);
        if (start != end) {
          setupMapper(seed);
        } else {
          setupMapper([]);
        }
        const result = await service.getChartData(settDefinition, new Date(start), new Date(end), granularity, size);
        expect(result.length).toEqual(expected);
        for (const point of result) {
          expect(seed.includes(point)).toBeTruthy();
        }
      },
    );
  });
});
