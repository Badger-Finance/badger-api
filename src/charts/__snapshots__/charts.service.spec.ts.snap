// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`charts.service loadVaultChartData returns requested vault chart data if exists 1`] = `
Array [
  HistoricVaultSnapshotModel {
    "address": "0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551",
    "apr": 37.36199709573286,
    "available": 93.04307958110472,
    "balance": 160912.73437961238,
    "block": 15298287,
    "boostWeight": 0,
    "harvestApr": 37.36199709573286,
    "id": "ethereum-0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551",
    "pricePerFullShare": 1,
    "strategy": Object {
      "address": "0x98Ca7AFa876f0e15494E76E92C5b3658cdE1Ffe1",
      "aumFee": 0,
      "performanceFee": 0,
      "strategistFee": 0,
      "withdrawFee": 10,
    },
    "strategyBalance": 160819.69130003126,
    "timestamp": 1660223160457,
    "totalSupply": 160912.73437961238,
    "value": 1116121.18,
    "yieldApr": 37.36199709573286,
  },
  HistoricVaultSnapshotModel {
    "address": "0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551",
    "apr": 36.42043576354977,
    "available": 93.04307958110472,
    "balance": 160912.73437961238,
    "block": 15291696,
    "boostWeight": 0,
    "harvestApr": 36.42043576354977,
    "id": "ethereum-0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551",
    "pricePerFullShare": 1,
    "strategy": Object {
      "address": "0x98Ca7AFa876f0e15494E76E92C5b3658cdE1Ffe1",
      "aumFee": 0,
      "performanceFee": 0,
      "strategistFee": 0,
      "withdrawFee": 10,
    },
    "strategyBalance": 160819.69130003126,
    "timestamp": 1660223160457,
    "totalSupply": 160912.73437961238,
    "value": 1137078.69,
    "yieldApr": 36.42043576354977,
  },
  HistoricVaultSnapshotModel {
    "address": "0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551",
    "apr": 36.058699370488235,
    "available": 227.79837966012852,
    "balance": 159193.72047228398,
    "block": 15285176,
    "boostWeight": 0,
    "harvestApr": 36.058699370488235,
    "id": "ethereum-0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551",
    "pricePerFullShare": 1,
    "strategy": Object {
      "address": "0x98Ca7AFa876f0e15494E76E92C5b3658cdE1Ffe1",
      "aumFee": 0,
      "performanceFee": 0,
      "strategistFee": 0,
      "withdrawFee": 10,
    },
    "strategyBalance": 158965.92209262386,
    "timestamp": 1660223160457,
    "totalSupply": 159193.72047228398,
    "value": 1128118.19,
    "yieldApr": 36.058699370488235,
  },
  HistoricVaultSnapshotModel {
    "address": "0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551",
    "apr": 36.18786526089967,
    "available": 227.79837966012852,
    "balance": 159193.72047228398,
    "block": 15278533,
    "boostWeight": 0,
    "harvestApr": 0.006305366586500938,
    "id": "ethereum-0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551",
    "pricePerFullShare": 1,
    "strategy": Object {
      "address": "0x98Ca7AFa876f0e15494E76E92C5b3658cdE1Ffe1",
      "aumFee": 0,
      "performanceFee": 0,
      "strategistFee": 0,
      "withdrawFee": 10,
    },
    "strategyBalance": 158965.92209262386,
    "timestamp": 1660223160457,
    "totalSupply": 159193.72047228398,
    "value": 1097323.6,
    "yieldApr": 0.006305366586500938,
  },
]
`;
